---
title: Exploring Homoscedasticity in Gene Expression
author: Kieran Mace
date: '2018-02-01'
output:
  html_document:
    df_print: paged
---



<pre class="r"><code>library(DESeq2)</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:parallel&#39;:
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, cbind, colMeans,
##     colnames, colSums, do.call, duplicated, eval, evalq, Filter,
##     Find, get, grep, grepl, intersect, is.unsorted, lapply,
##     lengths, Map, mapply, match, mget, order, paste, pmax,
##     pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce,
##     rowMeans, rownames, rowSums, sapply, setdiff, sort, table,
##     tapply, union, unique, unsplit, which, which.max, which.min</code></pre>
<pre><code>## 
## Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     expand.grid</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: SummarizedExperiment</code></pre>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     &#39;browseVignettes()&#39;. To cite Bioconductor, see
##     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>## Loading required package: DelayedArray</code></pre>
<pre><code>## Loading required package: matrixStats</code></pre>
<pre><code>## 
## Attaching package: &#39;matrixStats&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:Biobase&#39;:
## 
##     anyMissing, rowMedians</code></pre>
<pre><code>## 
## Attaching package: &#39;DelayedArray&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:matrixStats&#39;:
## 
##     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     apply</code></pre>
<pre class="r"><code>dds &lt;- makeExampleDESeqDataSet(n=50000,m=2)
reads = counts(dds)
plot(log2(reads[,1]),log2(reads[,2]))</code></pre>
<p><img src="/post/Explaining_Homoscedasticity/Exploring_Homoscedasticity_files/figure-html/unnamed-chunk-1-1.png" width="672" /></p>
<pre class="r"><code>reads_mean = apply(reads,1,mean)
reads_sd = apply(reads,1, sd)
plot(reads_mean, reads_sd)</code></pre>
<p><img src="/post/Explaining_Homoscedasticity/Exploring_Homoscedasticity_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
<p>Now lets normalize library size and esimate dispersion:</p>
<pre class="r"><code>dds = estimateSizeFactors(dds)
estimateDispersions(dds)</code></pre>
<pre><code>## Warning in checkForExperimentalReplicates(object, modelMatrix): same number of samples and coefficients to fit,
##   estimating dispersion by treating samples as replicates.
##   please read the ?DESeq section on &#39;Experiments without replicates&#39;.
##   in summary: this analysis only potentially useful for data exploration,
##   accurate differential expression analysis requires replication</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## class: DESeqDataSet 
## dim: 50000 2 
## metadata(1): version
## assays(2): counts mu
## rownames(50000): gene1 gene2 ... gene49999 gene50000
## rowData names(12): trueIntercept trueBeta ... dispOutlier dispMAP
## colnames(2): sample1 sample2
## colData names(2): condition sizeFactor</code></pre>
<pre class="r"><code>ntd = assay(normTransform(dds)) # regular log2 with a psudocount
rld = assay(rlog(dds)) # doing the same thing as log2, but not allowing low counts to explode </code></pre>
<pre class="r"><code>library(&quot;vsn&quot;)
meanSdPlot(ntd, ranks=F)</code></pre>
<p><img src="/post/Explaining_Homoscedasticity/Exploring_Homoscedasticity_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<pre class="r"><code>meanSdPlot(rld, ranks=F)</code></pre>
<p><img src="/post/Explaining_Homoscedasticity/Exploring_Homoscedasticity_files/figure-html/unnamed-chunk-4-2.png" width="672" /></p>
<p>So here we can see that we have shrunk the sd at low levels</p>
<pre class="r"><code>plot(rld[,1],rld[,2])</code></pre>
<p><img src="/post/Explaining_Homoscedasticity/Exploring_Homoscedasticity_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>now lets do it ourselves:</p>
<pre class="r"><code>ntd_mean = apply(ntd,1,mean)
ntd_sd = apply(ntd,1,sd)
plot(ntd_mean,ntd_sd)</code></pre>
<p><img src="/post/Explaining_Homoscedasticity/Exploring_Homoscedasticity_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<pre class="r"><code>library(mgcv)</code></pre>
<pre><code>## Loading required package: nlme</code></pre>
<pre><code>## 
## Attaching package: &#39;nlme&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     collapse</code></pre>
<pre><code>## This is mgcv 1.8-23. For overview type &#39;help(&quot;mgcv-package&quot;)&#39;.</code></pre>
<pre class="r"><code>model &lt;- smooth.spline(ntd_mean, ntd_sd) # build the model
fit &lt;- predict( model , se = TRUE )$fit # estimated values
se &lt;- predict( model , se = TRUE)$se.fit # standard error

# Confidence interval:
lcl &lt;- fit - 1.96 * se
ucl &lt;- fit + 1.96 * se


fit &lt;- smooth.spline(ntd_mean, ntd_sd)      # smooth.spline fit
res &lt;- (fit$yin - fit$y)/(1-fit$lev)      # jackknife residuals
sigma &lt;- sqrt(var(res))                     # estimate sd

upper &lt;- fit$y + 2.0*sigma*sqrt(fit$lev)   # upper 95% conf. band
lower &lt;- fit$y - 2.0*sigma*sqrt(fit$lev) 

data = data.frame(estimated_mean = fit$x,
                  estimated_sd = fit$y,
                  lower_bound = lower,
                  upper_bound = upper)

library(ggplot2)

ggplot(data) +
       geom_line(aes(x = estimated_mean,
                     y = estimated_sd)) + 
       geom_ribbon(aes(x = estimated_mean,
                       ymax = upper_bound,
                       ymin = lower_bound), alpha=0.5)</code></pre>
<p><img src="/post/Explaining_Homoscedasticity/Exploring_Homoscedasticity_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
