<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Hugo 0.42.1" />


<title>Exploring Homoscedasticity in Gene Expression - Mutual Information</title>
<meta property="og:title" content="Exploring Homoscedasticity in Gene Expression - Mutual Information">



  








<link href='//cdn.bootcss.com/highlight.js/9.11.0/styles/github.min.css' rel='stylesheet' type='text/css' />



<link rel="stylesheet" href="/css/fonts.css" media="all">
<link rel="stylesheet" href="/css/main.css" media="all">



  </head>
  <body>
    <div class="wrapper">
      <header class="header">
        <nav class="nav">
  <a href="/" class="nav-logo">
    <img src="/images/logo.png"
         width="50"
         height="50"
         alt="Logo">
  </a>

  <ul class="nav-links">
    
    <li><a href="/about/">About</a></li>
    
    <li><a href="https://github.com/kmace">GitHub</a></li>
    
    <li><a href="/resume.pdf">Resume</a></li>
    
  </ul>
</nav>

      </header>


<main class="content" role="main">

  <article class="article">
    
    <span class="article-duration">3 min read</span>
    

    <h1 class="article-title">Exploring Homoscedasticity in Gene Expression</h1>

    
    <span class="article-date">2018/02/01</span>
    

    <div class="article-content">
      <pre class="r"><code>library(DESeq2)</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:parallel&#39;:
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, basename, cbind,
##     colMeans, colnames, colSums, dirname, do.call, duplicated,
##     eval, evalq, Filter, Find, get, grep, grepl, intersect,
##     is.unsorted, lapply, lengths, Map, mapply, match, mget, order,
##     paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind,
##     Reduce, rowMeans, rownames, rowSums, sapply, setdiff, sort,
##     table, tapply, union, unique, unsplit, which, which.max,
##     which.min</code></pre>
<pre><code>## 
## Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     expand.grid</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: SummarizedExperiment</code></pre>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     &#39;browseVignettes()&#39;. To cite Bioconductor, see
##     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>## Loading required package: DelayedArray</code></pre>
<pre><code>## Loading required package: matrixStats</code></pre>
<pre><code>## 
## Attaching package: &#39;matrixStats&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:Biobase&#39;:
## 
##     anyMissing, rowMedians</code></pre>
<pre><code>## Loading required package: BiocParallel</code></pre>
<pre><code>## 
## Attaching package: &#39;DelayedArray&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:matrixStats&#39;:
## 
##     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     aperm, apply</code></pre>
<pre class="r"><code>dds &lt;- makeExampleDESeqDataSet(n=50000,m=4)
reads = counts(dds)
plot(log2(reads[,1]),log2(reads[,2]))</code></pre>
<p><img src="/post/Explaining_Homoscedasticity/Exploring_Homoscedasticity_files/figure-html/unnamed-chunk-1-1.png" width="672" /></p>
<pre class="r"><code>reads_mean = apply(reads,1,mean)
reads_sd = apply(reads,1, sd)
plot(reads_mean, reads_sd)</code></pre>
<p><img src="/post/Explaining_Homoscedasticity/Exploring_Homoscedasticity_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
<p>Now lets normalize library size and esimate dispersion:</p>
<pre class="r"><code>dds = estimateSizeFactors(dds)
#estimateDispersions(dds)
ntd = assay(normTransform(dds)) # regular log2 with a psudocount
rld = assay(rlog(dds)) # doing the same thing as log2, but not allowing low counts to explode </code></pre>
<pre class="r"><code>library(&quot;vsn&quot;)
meanSdPlot(ntd, ranks=F)</code></pre>
<p><img src="/post/Explaining_Homoscedasticity/Exploring_Homoscedasticity_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<pre class="r"><code>meanSdPlot(rld, ranks=F)</code></pre>
<p><img src="/post/Explaining_Homoscedasticity/Exploring_Homoscedasticity_files/figure-html/unnamed-chunk-4-2.png" width="672" /></p>
<p>So here we can see that we have shrunk the sd at low levels</p>
<pre class="r"><code>plot(rld[,1],rld[,2])</code></pre>
<p><img src="/post/Explaining_Homoscedasticity/Exploring_Homoscedasticity_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>now lets do it ourselves:</p>
<pre class="r"><code>ntd_mean = apply(ntd,1,mean)
ntd_sd = apply(ntd,1,sd)
plot(ntd_mean,ntd_sd)</code></pre>
<p><img src="/post/Explaining_Homoscedasticity/Exploring_Homoscedasticity_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<pre class="r"><code>library(mgcv)</code></pre>
<pre><code>## Loading required package: nlme</code></pre>
<pre><code>## 
## Attaching package: &#39;nlme&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     collapse</code></pre>
<pre><code>## This is mgcv 1.8-28. For overview type &#39;help(&quot;mgcv-package&quot;)&#39;.</code></pre>
<pre class="r"><code>model &lt;- smooth.spline(ntd_mean, ntd_sd) # build the model
fit &lt;- predict( model , se = TRUE )$fit # estimated values
se &lt;- predict( model , se = TRUE)$se.fit # standard error

# Confidence interval:
lcl &lt;- fit - 1.96 * se
ucl &lt;- fit + 1.96 * se


fit &lt;- smooth.spline(ntd_mean, ntd_sd)      # smooth.spline fit
res &lt;- (fit$yin - fit$y)/(1-fit$lev)      # jackknife residuals
sigma &lt;- sqrt(var(res))                     # estimate sd

upper &lt;- fit$y + 2.0*sigma*sqrt(fit$lev)   # upper 95% conf. band
lower &lt;- fit$y - 2.0*sigma*sqrt(fit$lev) 

data = data.frame(estimated_mean = fit$x,
                  estimated_sd = fit$y,
                  lower_bound = lower,
                  upper_bound = upper)

library(ggplot2)

ggplot(data) +
       geom_line(aes(x = estimated_mean,
                     y = estimated_sd)) + 
       geom_ribbon(aes(x = estimated_mean,
                       ymax = upper_bound,
                       ymin = lower_bound), alpha=0.5)</code></pre>
<p><img src="/post/Explaining_Homoscedasticity/Exploring_Homoscedasticity_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>

    </div>
  </article>

  


</main>

      <footer class="footer">
        <ul class="footer-links">
          <li>
            <a href="/index.xml" type="application/rss+xml" target="_blank">RSS feed</a>
          </li>
          <li>
            <a href="https://gohugo.io/" class="footer-links-kudos">Made with <img src="/images/hugo-logo.png" width="22" height="22"></a>
          </li>
        </ul>
      </footer>

    </div>
    



<script src="//cdn.bootcss.com/highlight.js/9.11.0/highlight.min.js"></script>



<script src="//cdn.bootcss.com/highlight.js/9.11.0/languages/r.min.js"></script>
<script src="//cdn.bootcss.com/highlight.js/9.11.0/languages/yaml.min.js"></script>
<script>hljs.configure({languages: []}); hljs.initHighlightingOnLoad();</script>



    
<script src="/js/math-code.js"></script>
<script async src="//cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML"></script>


    
  </body>
</html>

